connectors/src/connectors/notion/temporal/workflows/deletion_crawl.ts

```diff
@@ -12,103 +12,83 @@ import { notionDeletionCrawlSignal } from "@connectors/connectors/notion/tempora
 import type { ModelId } from "@connectors/types";
 
 const {
-  checkResourceAndQueueRelated,
   batchDeleteResources,
   clearWorkflowCache,
+  deletionCrawlAddSignalsToRedis,
+  deletionCrawlProcessBatch,
+  completeDeletionCrawlRun,
 } = proxyActivities<typeof activities>({
   startToCloseTimeout: "10 minute",
 });
 
-const DELETE_BATCH_SIZE = 100;
+const DELETE_BATCH_SIZE = 50;
 
 /**
  * Signal-based deletion crawl workflow.
- * Receives signals for resources to check, maintains a "seen" cache to avoid duplicates,
+ * Receives signals for resources to check, maintains a "seen" cache in Redis to avoid duplicates,
  * and recursively checks parent/children relationships.
  * Auto-terminates after 5 minutes of inactivity.
  */
 export async function notionDeletionCrawlWorkflow({
   connectorId,
-  resourceQueue = [],
 }: {
   connectorId: ModelId;
-  resourceQueue?: NotionDeletionCrawlSignal[];
 }) {
-  const seen = new Set<string>();
   const topLevelWorkflowId = workflowInfo().workflowId;
 
-  // Clear workflow cache at start
+  // Clear workflow cache (pg) at start
   await clearWorkflowCache({
     connectorId,
     topLevelWorkflowId,
   });
 
-  const resourceKey = (resourceId: string, resourceType: string) =>
-    `${resourceType}:${resourceId}`;
-  const hasBeenSeen = (resourceId: string, resourceType: string): boolean => {
-    return seen.has(resourceKey(resourceId, resourceType));
-  };
+  // Pending signals from signal handler (in-memory only)
+  const pendingSignals: NotionDeletionCrawlSignal[] = [];
 
   setHandler(notionDeletionCrawlSignal, (signal: NotionDeletionCrawlSignal) => {
-    if (!hasBeenSeen(signal.resourceId, signal.resourceType)) {
-      resourceQueue.push(signal);
-    }
+    pendingSignals.push(signal);
   });
 
   for (;;) {
-    // Wait for a resource signal, but stop the workflow if no signals arrive for 5 minutes
-    if (!(await condition(() => resourceQueue.length > 0, "5 minutes"))) {
+    // Wait for signals to arrive, but stop the workflow if no signals arrive for 5 minutes
+    if (!(await condition(() => pendingSignals.length > 0, "5 minutes"))) {
+      // Normal exit: cleanup Redis keys
+      await completeDeletionCrawlRun({
+        connectorId,
+        workflowId: topLevelWorkflowId,
+      });
       return;
     }
 
-    while (resourceQueue.length > 0) {
-      const resource = resourceQueue.shift();
-      if (!resource) {
-        continue;
-      }
-
-      const key = resourceKey(resource.resourceId, resource.resourceType);
-
-      seen.add(key);
-
-      // check only direct children + parent
-      const discovered = await checkResourceAndQueueRelated({
+    if (pendingSignals.length > 0) {
```

Comment:
Comment here, e.g.
If we have pending signals, copy them all to the redis queue and reset the signal list